crypto/rand.go

```diff
@@ -2,109 +2,103 @@ package crypto
 
 import (
 	"crypto/rand"
+	"hash"
+	"io"
+	"math"
 	"math/big"
-	"sync/atomic"
+	"sync"
+	"unsafe"
 )
 
-var (
-	// Two counters is used in case the first overflows. The first is not at
-	// risk of overflowing, but we're paranoid.
-	counter     uint64
-	counter2    uint64
-	entropyBase Hash
-)
-
-// init will create an entropy pool that the RNG will draw from. On most
-// systems, hashing out of a preset RNG pool will be substantially faster than
-// using crypto/rand, which relies on syscalls.
-func init() {
-	// Use 64 bytes in case the first 32 aren't completely random.
-	base := make([]byte, 64)
-	_, err := rand.Read(base)
-	if err != nil {
-		panic("unable to set entropy for the crypto package")
-	}
-	entropyBase = HashObject(base)
+// A randReader produces random values via repeated hashing. The entropy field
+// is the concatenation of an initial seed and a 128-bit counter. Each time
+// the entropy is hashed, the counter is incremented.
+type randReader struct {
+	entropy [16 + HashSize]byte
+	h       hash.Hash
+	buf     [32]byte
+	mu      sync.Mutex
 }
 
-// RandBytes returns n bytes of random data.
-func RandBytes(n int) ([]byte, error) {
-	b := make([]byte, n)
-	for i := 0; i < n; i += HashSize {
-		// Fetch and update the counter values before executing the hash.
-		var newCounter, newCounter2 uint64
-		newCounter = atomic.AddUint64(&counter, 1)
-		if newCounter == 0 {
-			newCounter2 = atomic.AddUint64(&counter2, 1)
-		} else {
-			newCounter2 = atomic.LoadUint64(&counter2)
+// Read fills b with random data. It always returns len(b), nil.
+func (r *randReader) Read(b []byte) (int, error) {
+	r.mu.Lock()
+	n := 0
+	for n < len(b) {
+		// Increment counter.
+		*(*uint64)(unsafe.Pointer(&r.entropy[0]))++
+		if *(*uint64)(unsafe.Pointer(&r.entropy[0])) == 0 {
+			*(*uint64)(unsafe.Pointer(&r.entropy[8]))++
 		}
-		// Grab some entropy using the unique counter set.
-		entropy := HashAll(newCounter, newCounter2, entropyBase)
+		// Hash the counter + initial seed.
+		r.h.Reset()
+		r.h.Write(r.entropy[:])
+		r.h.Sum(r.buf[:0])
 
 		// Fill out 'b'.
-		copy(b[i:], entropy[:])
+		n += copy(b[n:], r.buf[:])
 	}
-	return b, nil
+	r.mu.Unlock()
```

Comment:
would prefer to see atomic operations here, there are situations where the rng is going to be used in parallel, and there's no reason that it shouldn't do speedups when you use it in parallel. This lock is held during all the hashing, meaning parallelism offers no advantages.

A single atomic uint64 is good enough, then when it gets up above a certain value (2^63 for example) you mix in more entropy and then reset it. That can happen safely in a separate goroutine.